packages/context/src/context.ts

```diff
@@ -46,6 +113,194 @@ export class Context {
     this.name = name || uuidv1();
   }
 
+  /**
+   * Wrap the debug statement so that it always print out the context name
+   * as the prefix
+   * @param args Arguments for the debug
+   */
+  // tslint:disable-next-line:no-any
+  private _debug(...args: any[]) {
+    /* istanbul ignore if */
+    if (!debug.enabled) return;
+    const formatter = args.shift();
+    if (typeof formatter === 'string') {
+      debug(`[%s] ${formatter}`, this.name, ...args);
+    } else {
+      debug('[%s] ', this.name, formatter, ...args);
+    }
+  }
+
+  /**
+   * Set up an internal listener to notify registered observers asynchronously
+   * upon `bind` and `unbind` events. This method will be called lazily when
+   * the first observer is added.
+   */
+  private setupEventHandlersIfNeeded() {
+    if (this.notificationQueue != null) return;
+
+    this.addParentEventListener('bind');
+    this.addParentEventListener('unbind');
+
+    // The following are two async functions. Returned promises are ignored as
+    // they are long-running background tasks.
+    this.startNotificationTask().catch(err => {
+      this.handleNotificationError(err);
+    });
+
+    let ctx = this._parent;
+    while (ctx) {
+      ctx.setupEventHandlersIfNeeded();
+      ctx = ctx._parent;
+    }
+  }
+
+  /**
+   * Add an event listener to its parent context so that this context will
+   * be notified of parent events, such as `bind` or `unbind`.
+   * @param event Event name
+   */
+  private addParentEventListener(event: string) {
+    if (this._parent == null) return;
+
+    // Keep track of parent event listeners so that we can remove them
+    this._parentEventListeners = this._parentEventListeners || new Map();
+    if (this._parentEventListeners.has(event)) return;
+
+    const parentEventListener = (
+      binding: Readonly<Binding<unknown>>,
+      context: Context,
+    ) => {
+      // Propagate the event to this context only if the binding key does not
+      // exist in this context. The parent binding is shadowed if there is a
+      // binding with the same key in this one.
+      if (this.contains(binding.key)) {
+        this._debug(
+          'Event %s %s is not re-emitted from %s to %s',
+          event,
+          binding.key,
+          context.name,
+          this.name,
+        );
+        return;
+      }
+      this._debug(
+        'Re-emitting %s %s from %s to %s',
+        event,
+        binding.key,
+        context.name,
+        this.name,
+      );
+      this.emit(event, binding, context);
+    };
+    this._parentEventListeners.set(event, parentEventListener);
+    // Listen on the parent context events
+    this._parent.on(event, parentEventListener);
+  }
+
+  /**
+   * Handle errors caught during the notification of observers
+   * @param err Error
+   */
+  private handleNotificationError(err: unknown) {
+    this._debug(err);
```

Comment:
I think this debug statement will be confusing to our users - it prints the error only but does not give any context where is the error coming from (a context-event listener).

I am proposing to remove this `_debug` statement and include the error in the debug statements on lines 215 and 220 below.